### ISSUE TITLE
Access to the current file from the `replacement` function

### ISSUE THREAD

**Issue by gitgrimbo** [eruizdechavez/grunt-string-replace#40]
Can a `replacement` function access the file-name/file-path of the file that is currently being processed?

E.g.

````js
    replacements: [{
        pattern: /<img src=["']([^"']*?)["']/ig,
        replacement: function(match, p1, offset, wholeString) {
            var newPath = derivePathFrom(CURRENT_SOURCE_FILE);
            return '<img src="' + newPath + '/' + p1 + '"';
        }
    }]
````

If not, can this be added?

**Comment by eruizdechavez** [comment]
@gitgrimbo it is a good idea and I figured out a way to do it.

Would you mind testing it with `npm install grunt-string-replace@beta`?

Your replacement function will receive 2 extra parameters at the end: `src` and `dest` with the corresponding values that grunt reports to the plugin.

```js
    replacements: [{
        pattern: /<img src=["']([^"']*?)["']/ig,
        replacement: function(match, p1, offset, wholeString, src, dest) {
            var newPath = derivePathFrom(CURRENT_SOURCE_FILE);
            return '<img src="' + newPath + '/' + p1 + '"';
        }
    }]
```

**Comment by gitgrimbo** [comment]
Hi, yes that seems to work, thanks.

FYI, here is an example of why I wanted this (could be useful documentation for someone else).

### Example

I'm using a [Hugo](https://gohugo.io/) generated site.

This is how my images look in the Markdown file, they are relative links (so I don't have any path info here, which is why I wanted a local file path to the image):

`![](20160331_221339.jpg)`

This gets translated to HTML by Hugo, e.g.:

`<img src="20160331_221339.jpg" alt="" />`

And I can use the new `grunt-string-replace` beta to transform this to:

`<a href=https://EXTERNAL_SITE/garden-solar-lights/ncp1402/20160331_221339.jpg><img src="/small-img/garden-solar-lights/ncp1402/20160331_221339.jpg" alt="" data-original-src="20160331_221339.jpg"></a>`

which has replaced the `<img src>` to a small version of the image (for smaller downloads), and this `<img>` is wrapped in a link to the larger image, which is held on a different server (but using the same file path structure).  In my case, the server was Dropbox.

### Grunt code

Helper function (*Ignore the reference to `grunt.task.current.curFile`, I hacked a new value here to be the current file in the absence of the new beta functionality*):

This function uses the [parse5](https://github.com/inikulin/parse5) library to parse the `<img>` snippet and alter its attributes.

````js
function replaceImgSrc(grunt, localImgRoot, linkedImgRoot) {
    return {
        pattern: /<img\s+[^>]*>/ig,
        replacement: function(match, offset, wholeString, src, dest) {
            var frag = parse5.parseFragment(match);
            var img = frag.childNodes[0];
            // Can only serialize fragments
            grunt.verbose.writeln('imgStr', parse5.serialize(frag));

            var attrs = nvpArrayToOb(img.attrs);

            // If we have already processed this file, use data-original-src
            // rather than the src we have already modified.
            // This makes the script idempotent.
            var originalImgSrc = attrs['data-original-src'] || attrs['src'];

            grunt.verbose.writeln('data-original-src', attrs['data-original-src'], 'src', attrs['src']);

            if (!isRelative(originalImgSrc)) {
                return match;
            }

            // Get the full path to Hugo's 'public' folder
            var cwd = process.cwd();
            var root = path.join(cwd, 'public');

            // Get a relative path to the current (relative to 'public')
            var curFile = grunt.task.current.curFile || src;
            grunt.verbose.writeln('curFile', curFile, grunt.task.current.curFile, src);

            var relFile = path.relative(root, curFile);
            var relPath = path.dirname(relFile).replace(/\\/ig, '/');

            // The local images live in the '/public/img' folder (see 'responsive_images' task)
            var localImgSrc = localImgRoot + '/' + relPath + '/' + originalImgSrc;

            // Use the local image for the blog post and remember the original src.
            nvpArraySet(img.attrs, 'src', localImgSrc);
            nvpArraySet(img.attrs, 'data-original-src', originalImgSrc);

            // Enclose in an anchor to the full image.
            var result = '<a href=' + linkedImgRoot + '/' + relPath + '/' + originalImgSrc + '>' + parse5.serialize(frag) + '</a>';
            return result;
        }
    };
}
````

Grunt config snippet:

````js
        'string-replace': {
            images: {
                files: [{
                    expand: true,
                    src: 'public/**/*.html',
                    dest: '.'
                }],
                options: {
                    replacements: [
                        replaceImgSrc(grunt, '/small-img', 'https://EXTERNAL_SITE')
                    ]
                }
            }
        },
````

**Comment by eruizdechavez** [comment]
@gitgrimbo nice to hear it works for you, and thanks for the example. Once I get this documented (in the next days) I'll push a new release to get it live for everyone.

**Comment by eruizdechavez** [comment]
This was released in v1.3.0
